File: dp/122.best-time-to-buy-and-sell-stock-ii.py
#
# @lc app=leetcode id=122 lang=python3
#
# [122] Best Time to Buy and Sell Stock II
#

# @lc code=start


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices:
            return 0

        k = len(prices)
        hold0, hold1 = 0, -prices[0]

        for i in range(1, k):
            hold0, hold1 = max(
                hold0, hold1+prices[i]), max(hold1, hold0-prices[i])

        return hold0
# @lc code=end

File: dp/122.Best-Time-to-Buy-and-Sell-Stock-II.go
package dp

func maxProfit(prices []int) int {
	n := len(prices)
	if n == 0 {
		return 0
	}

	sell, hold := 0, -prices[0]

	for i := 1; i < n; i++ {
		sell = max(sell, hold+prices[i])
		hold = max(hold, sell-prices[i])
	}

	return sell
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}
